File: algorithms/visualstudio/algorithms.javascript/DivideAndConquer/MergeSort.js
var javascriptAlgorithms;
(function (javascriptAlgorithms) {
    (function (divideAndConquer) {

        var mergeSort = (function() { 

            function mergeSort(mathHelpersLibrary) {
                this.mathHelpers = mathHelpersLibrary;
            }

            mergeSort.prototype.sort = function(list) {
                if (list.length < 2) {
                    return list;
                }

                var splitPoint = this.mathHelpers.floatToInt(list.length / 2);
                var list1 = list.slice(0, splitPoint);
                var list2 = list.slice(splitPoint);

                var sortedList1 = this.sort(list1);
                var sortedList2 = this.sort(list2);

                return this.merge(sortedList1, sortedList2);
            };

            mergeSort.prototype.merge = function(list1, list2) {
                var mergedList = new Array;

                var list1Index = 0;
                var list2Index = 0;

                while (list1Index < list1.length && list2Index < list2.length) {
                    if (list1[list1Index] < list2[list2Index]) {
                        mergedList.push(list1[list1Index]);
                        list1Index++;
                    } else {
                        mergedList.push(list2[list2Index]);
                        list2Index++;
                    }
                };

                if (list1.length == list1Index) {
                    var tailOfList2 = list2.slice(list2Index);
                    mergedList = mergedList.concat(tailOfList2);
                };

                if (list2.length == list2Index) {
                    var tailOfList1 = list1.slice(list1Index);
                    mergedList = mergedList.concat(tailOfList1);
                };

                return mergedList;
            };

            return mergeSort;
        })();

        divideAndConquer.mergeSort = mergeSort;
    })(javascriptAlgorithms.divideAndConquer || (javascriptAlgorithms.divideAndConquer = {}));
})(javascriptAlgorithms || (javascriptAlgorithms = {}));

File: algorithms/visualstudio/algorithms.csharp/DivideAndConquer/MergeSort.cs
using System.Collections.Generic;
using System.Linq;

namespace algorithms.csharp.DivideAndConquer
{
    public class MergeSort
    {
        private int _list1Index;
        private int _list2Index;
        
        public List<int> Merge(List<int> list1, List<int> list2)
        {
            _list1Index = 0;
            _list2Index = 0;
            
            var mergedList = CompareAndMergeLists(list1, list2);

            if (list1.Count == _list1Index)
            {
                var tailOfList2 = list2.Skip(_list2Index);
                mergedList.AddRange(tailOfList2);
            }

            if (list2.Count == _list2Index)
            {
                var tailOfList1 = list1.Skip(_list1Index);
                mergedList.AddRange(tailOfList1);
            }

            return mergedList;
        }

        private List<int> CompareAndMergeLists(List<int> list1, List<int> list2)
        {
            var mergedList = new List<int>();

            while (_list1Index < list1.Count && _list2Index < list2.Count)
            {
                var currentList1Element = list1.ElementAt(_list1Index);
                var currentList2Element = list2.ElementAt(_list2Index);

                if (currentList1Element < currentList2Element)
                {
                    mergedList.Add(currentList1Element);
                    _list1Index++;
                }
                else
                {
                    mergedList.Add(currentList2Element);
                    _list2Index++;
                }
            }

            return mergedList;
        }

        public List<int> Sort(List<int> list)
        {
            if (list.Count < 2)
            {
                return list;
            }

            var splitPoint = list.Count / 2;

            var list1 = list.Take(splitPoint).ToList();
            var list2 = list.Skip(splitPoint).ToList();

            var sortedList1 = Sort(list1);
            var sortedList2 = Sort(list2);

            return Merge(sortedList1, sortedList2);
        }
    }
}
